content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md

```diff
@@ -0,0 +1,359 @@
+
+The most straightforward use is on a local machine. We switch to a virtual
+environment, type `jupyter notebook` in our terminal, and head over to
+http://localhost:8888. We're good to go and ready to start prototyping!
+
+This usage has one major downside, however: we're limited by the specs of our
+local machine. While my laptop won't struggle training a classifier for the
+[Iris dataset](https://www.kaggle.com/datasets/uciml/iris), it will become a
+limiting factor when working on an alternative to
+[GPT-3](https://github.com/openai/gpt-3).
+
+The solution to bypassing the limitations of our local machine is to move our
+notebooks to the cloud. When running a notebook on AWS or Azure, we can
+provision as beefy an instance as needed for our project. However, the downside
+to cloud-hosted notebooks is that they're more challenging to get up and
+running. And any time spent setting up our development environment is time we
+can't spend on our actual ML project.
+
+"Now wait a minute," you might say, "hasn't this problem long been solved by
+SaaS solutions like
+[Google Colab](https://colab.research.google.com/?utm_source=scs-index) and
+[Amazon SageMaker Studio](https://aws.amazon.com/sagemaker/studio/)?"
+
+To which I'd reply: "To an extent: yes. But while they certainly have their
+merits, these solutions often have their own limitations which can become
+problematic."
+
+Here are three of the limitations imposed by SaaS solutions such as Colab and
+SageMaker:
+
+- **Cost:** they are expensive to use. You're not only paying for the hardware
+  but also for using the platform.
+- **Walled garden:** they tend to be baked into the vendor's ecosystem. It's
+  difficult and sometimes impossible to use infrastructure components from
+  different vendors in conjunction with each other.
+- **Hardware limitations:** they may still put a limit on the hardware available
+  to you. Colab, for example, is capped at 25GB of RAM and caps uptime at 12
+  hours.
+
+Iterative provides an alternative to these services in the form of
+[TPI (Terraform Provider Iterative)](https://github.com/iterative/terraform-provider-iterative).
+This [Terraform](https://www.terraform.io/) plugin lets you extend the
+capabilities of your laptop by connecting it to endless cloud resources. We can
+use TPI to launch a wide variety of tasks without needing to be a cloud expert.
+Launching Jupyter workspaces is one such an application.
+
+Here are four benefits to TPI that might pique your interest:
+
+- **Free and open:** TPI is entirely open source and free to use.
+- **Lower cloud costs:** TPI automatically cleans up unused instances. It also
+  allows you to use spot instances, reducing your per-hour costs.
+- **Modular:** TPI can be used with a variety of cloud vendors, meaning you can
+  use it regardless of what the rest of your stack looks like.
+- **Developer-first experience:** {{is this really a benefit for target audience
```

Comment:
Ya. Good point.  This is a go-to for the ML engineer set, but maybe not as much for this audience.  

It could be replaced by some/all of the benefits Casper outlined in his comments before:

>No CI/CD required: You don’t need CI/CD to work in the cloud, you can get started right away
>Simple one file setup: install terraform, craft your main.tf >and you are gewd to go
>Extend your laptop hardware with the cloud: "plug your cloud into your laptop" for easy access to those endless resources
>Portable: Share your main.tf and anyone can run the experiment. Also you if you have data/storage in the cloud already, you can use an instance/compute with the same cloud provider in the same cloud region to make access/egress faster & more secure (important for bug data & enterprise)
>builtin spot recovery: We handle instance rebooting and syncing data so you can easily take full advantage of spot instances

Reply:
Did another iteration on this with the target audience in mind. The upsides now mirror the downsides of Colab etc. as listed above. 😄